### PR TITLE
Add TypeScript type definition

### PR DESCRIPTION
**Pull request by ExE-Boss**
Part of: https://github.com/sindresorhus/meta/issues/9

review?(@BendingBender, @sindresorhus)

### CODE REVIEW

**Comment by sindresorhus** [comment]
Don’t use `declare`, add doc comments, type tests, and follow https://github.com/sindresorhus/typescript-definition-style-guide

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+		 * @param blue (0-65535)
+		 */
+		apple(r16: number, g16: number, b16: number): string;
+
+		/**
+		 * Grayscale color
+		 *
+		 * @param gray (0-100)
+		 */
+		gray(gray: number): string;
```
Comment (sindresorhus) [inline index.d.ts]: Even though we just pass through the color modifiers to `color-convert`, I don't see the point in supporting all these conversions. Let's limit it to what we support in `chalk`: https://github.com/chalk/chalk#256-and-truecolor-color-support

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
+import {KEYWORD} from 'color-convert/conversions';
+
+declare namespace ansiStyles {
+	interface ColorConvert {
+		/**
+		 * The RGB color space
+		 *
```
Comment (sindresorhus) [inline index.d.ts]: Don't prefix lines with `*`.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+		readonly bgBlueBright: CSIPair;
+		readonly bgCyanBright: CSIPair;
+		readonly bgMagentaBright: CSIPair;
+		readonly bgWhiteBright: CSIPair;
+	}
+}
+
+declare const ansiStyles: {
+	readonly modifier: ansiStyles.Modifier;
+	readonly color: ansiStyles.ForegroundColor & ansiStyles.ColorBase<'\u001B[39m'>;
```
Comment (sindresorhus) [inline index.d.ts]: Is there any point/benefit of specifying exactly what string it returns instead of just `string`?

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+		 * @param gray (0-100)
+		 */
+		gray(gray: number): string;
+	}
+
+	interface CSIPair {
+		/** The ANSI terminal Control Sequence for setting this style */
+		readonly open: string;
+
+		/** The ANSI terminal Control Sequence for resetting this style */
```
Comment (sindresorhus) [inline index.d.ts]: "resetting" is not the correct work. I would say `starting` and `ending`.

File: index.d.ts (hunk 1 above)
Comment (ExE-Boss) [reply index.d.ts]: Well, the TypeScript definition is supposed to be accurate, and `ansi‑styles` currently supports all of the `color‑convert` color spaces, [which is a documented feature in `ansi‑styles`](https://github.com/chalk/ansi-styles#256--16-million-truecolor-support).

Removing support for those from `ansi‑styles` will be a breaking change,

Comment (sindresorhus) [reply index.d.ts]: Yes, that's planned regardless: https://github.com/chalk/chalk/issues/360

File: index.d.ts (hunk 3 above)
Comment (ExE-Boss) [reply index.d.ts]: ¯\\\_(ツ)\_/¯

Comment (sindresorhus) [reply index.d.ts]: Let's go with just `string`.

File: index.d.ts [hunk 5]
```diff
@@ -1,122 +1,124 @@
-import {KEYWORD} from 'color-convert/conversions';
+import * as cssColors from 'color-name';
```
Comment (sindresorhus) [inline index.d.ts]: Why did you change to this dependency? I would prefer not to add more dependencies here, even if it's just for TS.

Comment (ExE-Boss) [reply index.d.ts]: Because `color‑name` is used by `color‑convert`.

Comment (sindresorhus) [reply index.d.ts]: Alright

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+		*/
+		hex(hex: string): string;
+
+		/**
+		@param keyword - A CSS color name.
+		*/
+		keyword(keyword: keyof typeof cssColors): string;
+
+		ansi(ansi: number): string;
+		ansi256(ansi: number): string;
```
Comment (sindresorhus) [inline index.d.ts]: These are missing comments. See https://github.com/chalk/chalk/commit/fb8e85ab875eb95b55448fd615daacf8c325cf44 and https://github.com/chalk/chalk/pull/369

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+	interface Modifier {
+		readonly bold: CSIPair;
+		readonly dim: CSIPair;
+		readonly hidden: CSIPair;
+		readonly inverse: CSIPair;
+		readonly italic: CSIPair;
+		readonly reset: CSIPair;
+		readonly strikethrough: CSIPair;
+		readonly underline: CSIPair;
```
Comment (sindresorhus) [inline index.d.ts]: These could use comments: https://github.com/chalk/chalk/blob/4de1841129cf3d0a1db7a5d6638402b7828e1731/index.d.ts#L274-L318

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+		readonly red: CSIPair;
+		readonly green: CSIPair;
+		readonly yellow: CSIPair;
+		readonly blue: CSIPair;
+		readonly cyan: CSIPair;
+		readonly magenta: CSIPair;
+		readonly white: CSIPair;
+
+		readonly gray: CSIPair;
+		readonly grey: CSIPair;
```
Comment (sindresorhus) [inline index.d.ts]: Comments: https://github.com/chalk/chalk/blob/4de1841129cf3d0a1db7a5d6638402b7828e1731/index.d.ts#L357-L365

**Comment by sindresorhus** [comment]
Can you fix https://github.com/chalk/ansi-styles/pull/55/files#r329321851 ?
